Cython/Compiler/ModuleNode.py

```diff
@@ -2336,6 +2432,149 @@ def generate_import_star(self, env, code):
+        module_state_clear = globalstate['module_state_clear']
+        module_state_traverse = globalstate['module_state_traverse']
+        for module in modules:
+            definition = module is env
+            for entry in env.c_class_entries:
+                if definition or entry.defined_in_pxd:
+                    module_state.putln("PyObject *%s;" % entry.type.typeobj_cname)
+                    module_state_defines.putln("#define %s %s->%s" % (
+                        entry.type.typeobj_cname,
+                        Naming.modulestateglobal_cname,
+                        entry.type.typeobj_cname))
+                    module_state_clear.putln("  Py_CLEAR(%s(m)->%s);" % (
+                        Naming.modulestate_cname,
+                        entry.type.typeobj_cname))
+                    module_state_traverse.putln("  Py_VISIT(%s(m)->%s);" % (
+                        Naming.modulestate_cname,
+                        entry.type.typeobj_cname))
+        module_state.putln('} %s;' % Naming.modulestate_cname)
+        module_state.putln('')
+        module_state.putln('#ifdef __cplusplus')
+        module_state.putln('namespace {')
+        module_state.putln('extern struct PyModuleDef %s;' % Naming.pymoduledef_cname)
+        module_state.putln('} /* anonymous namespace */')
+        module_state.putln('#else')
+        module_state.putln('static struct PyModuleDef %s;' % Naming.pymoduledef_cname)
+        module_state.putln('#endif')
+        module_state.putln('')
+        module_state.putln('#define %s(o) ((%s *)PyModule_GetState(o))' % (
+            Naming.modulestate_cname,
+            Naming.modulestate_cname))
+        module_state.putln('')
+        module_state.putln('#define %s (%s(PyState_FindModule(&%s)))' % (
+            Naming.modulestateglobal_cname,
+            Naming.modulestate_cname,
+            Naming.pymoduledef_cname))
+        module_state.putln('')
+        module_state.putln('#define %s (PyState_FindModule(&%s))' % (
+            env.module_cname,
+            Naming.pymoduledef_cname))
+        module_state.putln("#endif")
+        module_state_defines.putln("#endif")
+        module_state_clear.putln("  return 0;")
+        module_state_clear.putln("};")
+        module_state_clear.putln("#endif")
+        module_state_traverse.putln("  return 0;")
+        module_state_traverse.putln("};")
+        module_state_traverse.putln("#endif")
+
+    def generate_module_state_defines(self, env, code):
+        code.putln("#if CYTHON_COMPILING_IN_LIMITED_API")
+        code.putln('#define %s %s->%s' % (
+            Naming.builtins_cname,
+            Naming.modulestateglobal_cname,
+            Naming.builtins_cname))
+        code.putln('#define %s %s->%s' % (
+            Naming.cython_runtime_cname,
+            Naming.modulestateglobal_cname,
+            Naming.cython_runtime_cname))
+        code.putln('#define %s %s->%s' % (
+            Naming.empty_tuple,
```

Comment:
I doubt that it's really still faster to look up the reference from the module state here, than to just call the C-API function that returns a new reference to the constant objects (tuple, bytes, unicode). At least in CPython, that is. At least it avoids having to clean up reference counts, e.g. when the constant is only used in comparisons.

Reply:
Currently yeah, it's probably much faster to just create the object. Although I believe that if we get rid of the TLS lookup (i.e: Petr's work), we should be able to get this to be just as fast.

For now yeah, the current approach that I have here avoid cleaning up the temporary variables that we would otherwise create all over the place. So let's leave this as is? :)